### PR TITLE
Update multinomial_op logits invalid arguments check description

### PR DESCRIPTION
**Pull request by redwrasse**
- Updates `InvalidArgument` descriptions to reflect overflow checks exist on the logits dimensions independently (batch_size and  num_classes), rather than simply a check on the overall logits.shape.

- Also adds in the `absl::InvalidArgumentError` construct for error messages as suggested by the reviewer.

### CODE REVIEW

**Review by bishalbaaniya123 — APPROVED** [review]

File: tensorflow/core/kernels/multinomial_op.cc [hunk 1]
```diff
@@ -163,17 +163,21 @@ class MultinomialOp : public OpKernel {
...
-      OP_REQUIRES(ctx, static_cast<int>(dim) == dim,
-                  errors::InvalidArgument(
-                      "logits.shape = ", logits_t.shape().DebugString(),
-                      " too large for int"));
-    }
-    const int batch_size = static_cast<int>(logits_t.dim_size(0));
-    const int num_classes = static_cast<int>(logits_t.dim_size(1));
+                    "num_samples should be non-negative, got ", num_samples));
+
+    const int64_t dim_batch_size = logits_t.dim_size(0);
```
Comment (sagunb) [inline tensorflow/core/kernels/multinomial_op.cc:168]: nit: why not define batch_size here (doesn't seem there is any benefit to having a dim_batch_size and a batch_size?

(same for the dim_num_classes and num_classes).

**Review by sagunb — COMMENTED** [review]
Thank you for the changes! Some minor readability related comment.

File: tensorflow/core/kernels/multinomial_op.cc (hunk 1 above)
Comment (redwrasse) [reply tensorflow/core/kernels/multinomial_op.cc:168]: Good idea, I'll update the variable names.

Comment (redwrasse) [reply tensorflow/core/kernels/multinomial_op.cc]: Removed `dim_num_classes` and `dim_batch_size` names.

**Review by sagunb — APPROVED** [review]
Thank you for the changes!

**Review by penpornk — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/multinomial_op.cc [hunk 2]
```diff
@@ -163,17 +163,17 @@ class MultinomialOp : public OpKernel {
...
-                      "logits.shape = ", logits_t.shape().DebugString(),
-                      " too large for int"));
-    }
+                    "num_samples should be non-negative, got ", num_samples));
+
     const int batch_size = static_cast<int>(logits_t.dim_size(0));
     const int num_classes = static_cast<int>(logits_t.dim_size(1));
+
+    OP_REQUIRES(ctx, batch_size == logits_t.dim_size(0),
+                errors::InvalidArgument("batch_size cannot exceed max int"));
```
[the remark this answers is not in the document]
Comment (redwrasse) [reply tensorflow/core/kernels/multinomial_op.cc]: Happy to. I updated all error messages in `multinomial_op.cc` to use the `absl::InvalidArgumentError` construct.

**Comment by keerthanakadiri** [comment]
Hi @sagunb, Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sagunb, Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sagunb, Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk , Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk , Can you please review this PR ? Thank you!

**Review by sagunb — APPROVED** [review]

**Comment by yuvashrikarunakaran** [comment]
To incorporate the requested changes — specifically, updating the InvalidArgument descriptions to reflect that overflow checks are performed independently on the batch_size and num_classes dimensions of logits, and adding the use of absl::InvalidArgumentError for error messages — here’s an updated version of the relevant code:

#include "tensorflow/core/platform/errors.h"
#include "absl/strings/str_cat.h"
#include "tensorflow/core/framework/tensor.h"

namespace tensorflow {

// Function to check for overflow in logits dimensions
void CheckLogitsDimensions(const Tensor& logits) {
  const int batch_size = logits.dim_size(0);  // First dimension: batch size
  const int num_classes = logits.dim_size(1); // Second dimension: number of classes

  // Check if batch_size or num_classes exceed the safe limits to prevent overflow
  constexpr int kMaxBatchSize = 1 << 20; // Set an appropriate maximum batch size (example)
  constexpr int kMaxNumClasses = 1 << 16; // Set an appropriate maximum number of classes (example)

  // Check for batch_size overflow
  if (batch_size > kMaxBatchSize) {
    const string error_message = absl::StrCat(
        "Batch size overflow: batch_size = ", batch_size,
        " exceeds the maximum allowed batch size of ", kMaxBatchSize);
    LOG(ERROR) << error_message;
    throw absl::InvalidArgumentError(error_message);
  }

  // Check for num_classes overflow
  if (num_classes > kMaxNumClasses) {
    const string error_message = absl::StrCat(
        "Number of classes overflow: num_classes = ", num_classes,
        " exceeds the maximum allowed number of classes of ", kMaxNumClasses);
    LOG(ERROR) << error_message;
    throw absl::InvalidArgumentError(error_message);
  }
}

// Example usage in a function handling logits
void ProcessLogits(const Tensor& logits) {
  // Perform independent overflow checks for batch_size and num_classes
  CheckLogitsDimensions(logits);

  // Additional logic to process logits can go here...
}

}  // namespace tensorflow

File: tensorflow/core/kernels/multinomial_op.cc [hunk 3]
```diff
@@ -153,30 +153,30 @@ class MultinomialOp : public OpKernel {
   void DoCompute(OpKernelContext* ctx, const Tensor& logits_t,
                  const Tensor& num_samples_t, GuardedPhiloxRandom* generator) {
     OP_REQUIRES(ctx, TensorShapeUtils::IsMatrix(logits_t.shape()),
-                errors::InvalidArgument("logits should be a matrix, got shape ",
-                                        logits_t.shape().DebugString()));
+                absl::InvalidArgumentError(absl::StrCat("logits should be a matrix, got shape ",
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/multinomial_op.cc:156]: Why did you change this?

**Review by mihaimaruseac — DISMISSED** [review]

**Review by mihaimaruseac — APPROVED** [review]
